lib/utils.ts

```diff
@@ -111,3 +112,18 @@ export function promiseWithTimeout<T>(ms: number, promise: Promise<T>): Promise<
     });
   });
 }
+
+
+/** Finds an exe file in the package assuming it is placed under `rootPath/platform-arch/exe`
+ * For example on Windows x64, if the `exeName` is `serve-d`, it returns the absolute path to `./bin/win32-x64/exeName.exe`
+ * @param exeName name of the exe file
+ * @param rootPath the path of the folder of the exe file. Defaults to 'join("bin", `${process.platform}-${process.arch}`)'
+ * @param exeExtention the extention of the exe file. Defaults to `process.platform === "win32" ? ".exe" : ""`
+ */
+export function getExePath(
+ exeName: string,
+ rootPath = join("bin", `${process.platform}-${process.arch}`),
+ exeExtention = process.platform === "win32" ? ".exe" : ""
+): string {
+ return resolve(join(rootPath, `${exeName}${exeExtention}`));
+}
```

Comment:
What if the file has the same executable for multiple platforms? would they have to copy the executable into multiple folders?

This utility function doesn't seem very useful since it is abstracting a one liner that is much more configurable.

Reply:
If the exe is the same, they can just pass it directly to spawn.

Reply:
> This is useful for native exe files, in which the exe is different for different for each platform and arch.

so only useful for very specific situation. Doesn't seem like a good candidate for a utility function.

Reply:
This is not "over specific". Rust, D, Go, C++, C, etc all have a similar situation.

Reply:
> In contrast, assuming that everything is a Node LSP is "over specific".

Nobody assumes that. It is just that writing `cp.spawn(args)` is just as easy as `super.spawn(args)` so it doesn't seem to make sense to abstract it.

Reply:
Well, I was stuck on this for two weeks. I thought that `spawnChildNode` is the way to use it for all the LSP stuff. Until I realized that is far from the truth.

Having some structure makes it easier to scale this and get reproducible results at scale. Not having anything considered gives the wrong signal.


For the same reason, that's why everyone uses `prebuildify`. This unified structure has made it easier to provide native addons at scale.  If you want to look at the issue from your point of view, `prebuildify` simply calls `prebuild` and copies the binaries to a predefined folder.

Reply:
Having the documentation on how to use an executable file would be good but if we want to abstract it for the `prebuildify` workflow that would be fine.

something like:

```js
super.spawnPrebuildify(
      'serve-d',
      args,
      opts
    )
```

anyone who wants more customizability can just us `cp.spawn`

Reply:
The current implementation I have chosen covers all the cases:
- native exe by using `getExePath` (useful for native languages such as D, C++)
- manual management
We already had Nodejs spawning which in combination with the above covers all the cases.

I don't want to become religious, but I think this provides a clean way to simplify the ide packages we already have. Many of the ide packages (e.g. ide-python) do manual error management for the spawned process because of the wrong signal this missing method had given, even though we already manage it internally!

Reply:
I'm not totally against abstracting `spawn` to make the "usual" process easier. It is mostly `getExePath` that seems like a waist of an abstraction.

Reply:
Here, I am in favor of providing small functions to allow the developers to compose them as they wish. This is cleaner to me in this specific case. 

But if you really want, I can merge `getExePath` into the `spawn` method. In my opinion, the interface is already clean, and merging `getExePath` can make it hard to use.

Reply:
I found a way to extend the functionality of `getExePath` to the cases in which the name is on the `PATH`

Reply:
> merging  `getExePath` can make it hard to use.

How can it make it hard to use? if they use prebuildify (or the same directory structure) they should be able to use the default configuration. (Which I'm assuming is most language servers.)

And if they do not use the same directory structure for the binaries it is just as simple to use `cp.spawn(serverExe, ...)` as it is to use `super.spawn(...)`

Esentially if the package uses the "usual" structure we make it easier. If they don't then it is the same as before.

Reply:
> Here, I am in favor of providing small functions to allow the developers to compose them as they wish.

For a general library like underscore, I agree, but this library is for very specific lsp atom packages.

Reply:
> I found a way to extend the functionality of `getExePath` to the cases in which the name is on the `PATH`

Shouldn't the package that is using `getExePath` know whether the exe is provided or it should return the `PATH`?

Reply:
> Shouldn't the package that is using getExePath know whether the exe is provided or it should be in the PATH?

Certainly if they bundle it with the package, but in another case, the file might not exist there. For example, you can say generally Python is installed in `~/usr/bin`, but it might not be there, and so you can use the exe on the PATH.

Reply:
Ya so if they know python is not bundled why would they use `exePath = getExePath("python")` instead of `exePath = "python"`?

Reply:
`getExePath` looks like a solution without a problem.

Reply:
Is `getExePath`  useful to lsp packages somewhere other than starting the server? If not, the code should be merged into the `spawn` function.

Reply:
Merged them in https://github.com/atom-community/atom-languageclient/pull/124/commits/a5c03a572bacd1a12eca83fbced80be942cc5ed6